File: Cell.cpp
#include "Cell.hpp"

std::ostream & operator<<(std::ostream& out, Cell c) {
	switch (c) {
		case Cell::EMPTY:
			out << ".";
			break;
		case Cell::SHIP:
			out << "#";
			break;
		case Cell::MISS:
			out << "-";
			break;
		case Cell::HIT:
			out << "X";
			break;
		default:
			std::cerr << "Unknown state" << std::endl;
	}
	return out;
}

std::istream& operator>>(std::istream& in, Cell& c) {
	char x;
	in >> x;
	switch (x) {
		case '.':
		case '0':
			c = Cell::EMPTY;
			break;
		case '#':
		case '1':
			c = Cell::SHIP;
			break;
		case '-':
			c = Cell::MISS;
			break;
		case 'X':
			c = Cell::HIT;
			break;
		default:
			std::cerr << "Unknown Cell value: " << x << std::endl;
	}
	return in;
}

File: cell.go
package main

import (
	"fmt"
	"io"
)

type cell int

const (
	EMPTY cell = 1
	SHIP  cell = 2
	MISS  cell = 3
	HIT   cell = 4
	KILL  cell = 5
)

func RuneToCell(c byte) cell {
	switch c {
	case '.', '0':
		return EMPTY
	case '#', '1':
		return SHIP
	case '-':
		return MISS
	case 'X':
		return HIT
	default:
		panic(fmt.Errorf("Unknown char: %c", c))
	}

}

func ReadCell(r io.Reader) cell {
	buf := make([]byte, 1)
	if _, err := r.Read(buf); err != nil {
		panic(err)
	}
	switch buf[0] {
	case '.', '0':
		return EMPTY
	case '#', '1':
		return SHIP
	case '-':
		return MISS
	case 'X':
		return HIT
	default:
		panic(fmt.Errorf("Unknown char: %c", buf[0]))
	}
}
